infra/build/functions/sync/main.py

```diff
@@ -35,41 +46,138 @@ class GitAuth(ndb.Model):
+      },
+      'schedule': project.schedule
+  }
+
+  update_mask = {'schedule': schedule}
+  cloud_scheduler_client.update(job, update_mask)
+
+
+def sync_projects(cloud_scheduler_client, projects):
   """Sync projects with cloud datastore."""
   project_query = Project.query()
-  projects_to_remove = [
-      project.key for project in project_query if project.name not in projects
-  ]
-
-  ndb.delete_multi(projects_to_remove)
+  for project in project_query:
+    if project.name not in projects:
+      try:
+        delete_scheduler(cloud_scheduler_client, project.name)
+        project.key.delete()
+      except exceptions.GoogleAPICallError as error:
+        logging.error('Scheduler deletion for %s failed with %s', project.name,
+                      error)
 
   existing_projects = {project.name for project in project_query}
 
-  new_projects = [
-      Project(name=project)
-      for project in projects
-      if project not in existing_projects
-  ]
-  ndb.put_multi(new_projects)
+  for project_name in projects:
+    if project_name not in existing_projects:
+      try:
+        create_scheduler(cloud_scheduler_client, project_name,
+                         projects[project_name])
+        Project(name=project_name, schedule=projects[project_name]).put()
+      except exceptions.GoogleAPICallError as error:
+        logging.error('Scheduler creation for %s failed with %s', project_name,
+                      error)
+
+  for project in project_query:
+    if project.name in projects and project.schedule != projects[project.name]:
+      try:
+        update_scheduler(cloud_scheduler_client, project,
+                         projects[project.name])
+        project.schedule = projects[project.name]
+        project.put()
+      except exceptions.GoogleAPICallError as error:
+        logging.error('Updating scheduler for %s failed with %s', project.name,
+                      error)
+
+
+def _has_docker_file(project_contents):
+  """Checks if project has a Dockerfile."""
+  return any(
+      content_file.name == 'Dockerfile' for content_file in project_contents)
 
 
-def _has_docker_file(repo, project_path):
-  """Checks if project has a Dockerfile."""
-  return any(content_file.name == 'Dockerfile'
-             for content_file in repo.get_contents(project_path))
+def get_schedule(project_contents):
+  """Checks for schedule parameter in yaml file else uses DEFAULT_SCHEDULE."""
+  for content_file in project_contents:
+    if content_file.name != 'project.yaml':
+      continue
+    yaml_str = content_file.decoded_content.decode('utf-8')
+    project_yaml = yaml.safe_load(yaml_str)
+    if 'schedule' in project_yaml:
+      try:
+        times_per_day = int(project_yaml['schedule'])
+        if times_per_day not in range(1, 5):
+          raise ProjectYamlError('Parameter not in range [1-4]')
+      except ValueError:
+        raise ProjectYamlError('Parameter not an integer')
```

Comment:
Put 4 in a global "MAX_BUILDS_PER_DAY". Then use MAX_BUILDS_PER_DAY + 1 in the range() call.

project_yaml['schedule'] will also raise a KeyError if schedule is not specified. 

The structure here looks a bit complicated. Let's simplify things by doing

```python
  times_per_day = project_yaml.get('schedule', DEFAULT_BUILDS_PER_DAY)
  if (not isinstance(times_per_day, int) or
      times_per_day not in range(1, MAX_BUILDS_PER_DAY + 1)):
    raise ProjectYamlError('schedule not an integer in the range [1-4]')

  ...
```

You can also get rid of the `if 'schedule' in project_yaml` part and the `DEFAULT_SCHEDULE` completely, and just handle everything in the one case.